File: apps/api/app/domain/workflow.py
"""Workflow 领域模型。"""

from dataclasses import dataclass, field
from datetime import datetime

from apps.api.app.domain.identity import utc_now


@dataclass(slots=True)
class Workflow:
    """工作流元信息。"""

    # workflow_id 是工作流唯一标识。
    workflow_id: str

    # org_id 是工作流所属组织。
    org_id: str

    # agent_id 是工作流绑定的 Agent。
    agent_id: str

    # name 是工作流名称。
    name: str

    # description 是工作流说明。
    description: str

    # draft_definition 保存当前草稿 DSL。
    draft_definition: dict[str, object]

    # published_version_id 保存当前发布版本 ID。
    published_version_id: str | None = None

    # created_by 是创建者用户 ID。
    created_by: str = ""

    # created_at 是创建时间。
    created_at: datetime = field(default_factory=utc_now)

    # updated_at 是更新时间。
    updated_at: datetime = field(default_factory=utc_now)


@dataclass(slots=True)
class WorkflowVersion:
    """工作流发布版本。"""

    # version_id 是版本唯一标识。
    version_id: str

    # workflow_id 是所属工作流。
    workflow_id: str

    # org_id 是所属组织。
    org_id: str

    # version_number 是递增版本号。
    version_number: int

    # definition 是发布时冻结的 DSL。
    definition: dict[str, object]

    # created_by 是发布者用户 ID。
    created_by: str

    # created_at 是发布时间。
    created_at: datetime = field(default_factory=utc_now)

File: apps/web/types/workflow.ts
/** Workflow types. */

export interface WorkflowItem {
  workflow_id: string;
  agent_id: string;
  name: string;
  description: string;
  draft_definition: WorkflowDefinition;
  published_version_id: string | null;
}

export interface WorkflowVersion {
  version_id: string;
  workflow_id: string;
  org_id: string;
  version_number: number;
  definition: WorkflowDefinition;
  release_note: string;
  created_by: string;
  created_at: string;
}

export interface WorkflowValidationResult {
  valid: boolean;
  errors: string[];
}

export interface WorkflowRun {
  run_id: string;
  workflow_id: string;
  version_id: string;
  agent_id: string;
  input_data: Record<string, unknown>;
  status: string;
  output_data: Record<string, unknown>;
  error_message: string;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
  updated_at: string | null;
}

export interface NodeRun {
  node_run_id: string;
  node_id: string;
  node_type: string;
  status: string;
  input_data: Record<string, unknown>;
  output_data: Record<string, unknown>;
  error_message: string;
  elapsed_ms: number;
  sequence?: number;
}

export interface WorkflowDefinition {
  version: string;
  nodes: WorkflowNodeConfig[];
  edges: WorkflowEdgeConfig[];
  /** Optional deterministic guardrails for one workflow run, never a billing policy. */
  execution_limits?: WorkflowExecutionLimits;
}

export interface WorkflowExecutionLimits {
  max_steps?: number;
  max_llm_calls?: number;
}

/**
 * A curated starting point for a new workflow.  Templates deliberately carry
 * only workflow DSL: they never include tenant-specific provider credentials,
 * knowledge-base IDs, or tool authorisations.
 */
export interface WorkflowTemplate {
  id: string;
  name: string;
  description: string;
  category: string;
  suggestedName: string;
  setup: string[];
  definition: WorkflowDefinition;
}

export interface WorkflowNodeConfig {
  id: string;
  type: string;
  config: Record<string, unknown>;
  /** Canvas coordinates. Persisted as editor metadata; the backend ignores it. */
  position?: { x: number; y: number };
}

export interface WorkflowEdgeConfig {
  source: string;
  target: string;
  /** Condition nodes route through one explicitly named output branch. */
  branch?: "true" | "false";
}

export interface CustomNodeData extends Record<string, unknown> {
  label: string;
  description?: string;
  capability?: "executable" | "schema";
  config?: Record<string, unknown>;
  /** Latest run status for this node, mapped from node runs. */
  runStatus?: string;
  /** Opens the quick-add menu anchored to this node (Dify-style "+" affordance). */
  onQuickAdd?: (nodeId: string, branch?: "true" | "false") => void;
}

export const CUSTOM_NODE_TYPES = {
  START: "start",
  END: "end",
  LLM: "llm",
  RAG: "rag",
  TOOL: "tool",
  CONDITION: "condition",
  HTTP: "http",
  CODE: "code",
  VARIABLE: "variable",
  TEMPLATE: "template",
  HUMAN: "human",
} as const;

export type CustomNodeType = (typeof CUSTOM_NODE_TYPES)[keyof typeof CUSTOM_NODE_TYPES];

export interface CreateWorkflowRequest {
  actor_user_id: string;
  agent_id: string;
  name: string;
  description: string;
  draft_definition: WorkflowDefinition;
}
